presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java

```diff
@@ -189,20 +192,61 @@ public void testArrayToJson()
+                new ArrayType(INTEGER),
+                asList(1, 0, 12345678, 12345679, 12345678, null));
+        assertFunction("CAST(JSON '[true, false, 1234567891234567, 1234567891234567.8, \"1234567891234567\", null]' AS ARRAY<BIGINT>)",
+                new ArrayType(BIGINT),
+                asList(1L, 0L, 1234567891234567L, 1234567891234568L, 1234567891234567L, null));
+
+        // real, double, decimal
+        assertFunction("CAST(JSON '[true, false, 12345, 12345.67, \"3.14\", \"NaN\", \"Infinity\", \"-Infinity\", null]' AS ARRAY<REAL>)",
+                new ArrayType(REAL),
+                asList(1.0f, 0.0f, 12345.0f, 12345.67f, 3.14f, Float.NaN, Float.POSITIVE_INFINITY, Float.NEGATIVE_INFINITY, null));
+        assertFunction("CAST(JSON '[true, false, 1234567890, 1234567890.1, \"3.14\", \"NaN\", \"Infinity\", \"-Infinity\", null]' AS ARRAY<DOUBLE>)",
+                new ArrayType(DOUBLE),
+                asList(1.0, 0.0, 1234567890.0, 1234567890.1, 3.14, Double.NaN, Double.POSITIVE_INFINITY, Double.NEGATIVE_INFINITY, null));
+        assertFunction("CAST(JSON '[true, false, 128, 123.456, \"3.14\", null]' AS ARRAY<DECIMAL(10, 5)>)",
+                new ArrayType(createDecimalType(10, 5)),
+                asList(decimal("1.00000"), decimal("0.00000"), decimal("128.00000"), decimal("123.45600"), decimal("3.14000"), null));
+        assertFunction("CAST(JSON '[true, false, 128, 12345678.12345678, \"3.14\", null]' AS ARRAY<DECIMAL(38, 8)>)",
+                new ArrayType(createDecimalType(38, 8)),
+                asList(decimal("1.00000000"), decimal("0.00000000"), decimal("128.00000000"), decimal("12345678.12345678"), decimal("3.14000000"), null));
+
+        // varchar, json
+        assertFunction("CAST(JSON '[true, false, 12, 12.3, \"puppies\", \"kittens\", \"null\", \"\", null]' AS ARRAY<VARCHAR>)",
+                new ArrayType(VARCHAR),
+                asList("true", "false", "12", "12.3", "puppies", "kittens", "null", "", null));
+        assertFunction("CAST(JSON '[5, 3.14, [1, 2, 3], \"e\", {\"a\": \"b\"}, null, \"null\", [null]]' AS ARRAY<JSON>)",
+                new ArrayType(JSON),
+                ImmutableList.of("5", "3.14", "[1,2,3]", "\"e\"", "{\"a\":\"b\"}", "null", "\"null\"", "[null]"));
+
+        // nested array/map
+        assertFunction("CAST(JSON '[[1, 2], [3, null], [], [null, null], null]' AS ARRAY<ARRAY<BIGINT>>)",
+                new ArrayType(new ArrayType(BIGINT)),
+                asList(asList(1L, 2L), asList(3L, null), asList(), asList(null, null), null));
+
+        // invalid cast
         assertInvalidCast("CAST(JSON '[1, null, 3]' AS ARRAY<TIMESTAMP>)");
         assertInvalidCast("CAST(JSON '[1, null, 3]' AS ARRAY<ARRAY<TIMESTAMP>>)");
-        assertInvalidCast("CAST(JSON '[1, 2, 3]' AS ARRAY<BOOLEAN>)");
         assertInvalidCast("CAST(JSON '[\"puppies\", \"kittens\"]' AS ARRAY<BIGINT>)");
```

Comment:
Make sure every `checkCondition` is covered in test. Please assert the error message when you assert invalid casts.